### DISCUSSION TITLE
Impose rotation in the structural mechanic solver

### DISCUSSION THREAD

**Discussion by MicheleBarucca97** [KratosMultiphysics/Kratos#10822] [Q&A]
Good morning everyone,

I am analyzing the case of a 3D wing to which I want to impose a rotation, but I have big problems with the convergence. For sure I miss something, can someone help me? I attach the non convergence of my problem:

ResidualBasedNewtonRaphsonStrategy: Setup Dofs Time: 0.00795328
ResidualBasedNewtonRaphsonStrategy: Setup System Time: 3.6245e-05
ResidualBasedNewtonRaphsonStrategy: System Matrix Resize Time: 0.0469448
ResidualBasedNewtonRaphsonStrategy: System Construction Time: 0.0550132
SolvingStrategy:  MESH MOVED 
Coupling iteration: 1 
ResidualBasedBlockBuilderAndSolver: Build time: 0.113181
ResidualBasedBlockBuilderAndSolver: System solve time: 0.817436
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 4.53786e+14;  Norm = 7.0705e+06
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 6.1471e+14;  Norm = 2510.61
ResidualBasedBlockBuilderAndSolver: Build time: 0.112322
ResidualBasedBlockBuilderAndSolver: System solve time: 0.800667
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 1.28651e+18;  Norm = 2.00453e+10
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 3.99281e+16;  Norm = 163075
ResidualBasedBlockBuilderAndSolver: Build time: 0.112007
ResidualBasedBlockBuilderAndSolver: System solve time: 0.805443
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 8.87695e+21;  Norm = 1.38313e+14
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 3.97375e+18;  Norm = 1.62297e+07
ResidualBasedBlockBuilderAndSolver: Build time: 0.116749
ResidualBasedBlockBuilderAndSolver: System solve time: 0.831219
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 2.13007e+24;  Norm = 3.3189e+16
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 2.73153e+19;  Norm = 1.11562e+08
ResidualBasedBlockBuilderAndSolver: Build time: 0.1179
ResidualBasedBlockBuilderAndSolver: System solve time: 0.843655
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 3.63685e+24;  Norm = 5.66662e+16
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 3.01394e+20;  Norm = 1.23096e+09
ResidualBasedBlockBuilderAndSolver: Build time: 0.139891
ResidualBasedBlockBuilderAndSolver: System solve time: 0.980882
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 3.37048e+24;  Norm = 5.2516e+16
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 2.8472e+20;  Norm = 1.16286e+09
ResidualBasedBlockBuilderAndSolver: Build time: 0.139516
ResidualBasedBlockBuilderAndSolver: System solve time: 0.928764
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 5.00949e+24;  Norm = 7.80537e+16
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 3.95498e+20;  Norm = 1.6153e+09
ResidualBasedBlockBuilderAndSolver: Build time: 0.118032
ResidualBasedBlockBuilderAndSolver: System solve time: 0.995054
SolvingStrategy:  MESH MOVED 
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL DISPLACEMENT CRITERION :: Ratio = 4.86463e+24;  Norm = 7.57965e+16
ResidualDisplacementAndOtherDoFCriteria: RESIDUAL ROTATION CRITERION :: Ratio = 3.87464e+20;  Norm = 1.58249e+09
ResidualBasedBlockBuilderAndSolver: Build time: 0.125233
ResidualBasedBlockBuilderAndSolver: System solve time: 0.984613
SolvingStrategy:  MESH MOVED 

Thanks in advance for the help

**Comment by loumalouomega** [comment]
How do you impose the rotation?, can you share the example?

**Reply by MicheleBarucca97** [reply]
Thanks for the answer @loumalouomega . Yes, my project parameters file is the following, I specify that I am doing an FSI simulation with other softwares. Also the rotation should be in degree or radiant? Anyway the file is the following:

{
  "problem_data": {
    "problem_name": "wing",
    "parallel_type": "OpenMP",
    "start_time": 0.0,
    "end_time": 1000000000000000.0,
    "echo_level": 1
  },
  "solver_settings": {
    "model_part_name": "Structure",
    "domain_size": 3,
    "solver_type": "dynamic",
    "time_integration_method": "implicit",
    "scheme_type": "bossak",
    "echo_level": 1,
    "analysis_type": "non_linear",
    "model_import_settings": {
      "input_type": "mdpa",
      "input_filename": "wingTrasl"
    },
    "material_import_settings": {
      "materials_filename": "StructuralMaterials.json"
    },
    "time_stepping": {
      "time_step": 0.000005
    },
    "line_search": false,
    "convergence_criterion": "residual_criterion",
    "displacement_relative_tolerance": 0.0001,
    "displacement_absolute_tolerance": 1e-08,
    "residual_relative_tolerance": 1e-03,
    "residual_absolute_tolerance": 1e-08,
    "max_iteration": 500,
    "move_mesh_flag": true,
    "linear_solver_settings": {
      "solver_type": "skyline_lu_factorization",
      "scaling": false,
      "verbosity": 0
    },
    "rotation_dofs": true
  },
  "processes": {
    "constraints_process_list": [
      {
        "python_module": "assign_vector_variable_process",
        "kratos_module": "KratosMultiphysics",
        "Parameters": {
          "mesh_id": 0,
          "model_part_name": "Structure.ROTATION_root",
          "variable_name": "ROTATION",
          "constrained": [
            true,
            true,
            true
          ],
          "value": [
            0.0,
            0.0,
            "((asin(0.8661)/asin(0.01))*asin(0.01*cos(256.224*t)))*pi/acos(-1)"
          ]
        }
      },
      {
        "python_module": "assign_vector_variable_process",
        "kratos_module": "KratosMultiphysics",
        "Parameters": {
          "mesh_id": 0,
          "model_part_name": "Structure.DISPLACEMENT_root",
          "variable_name": "DISPLACEMENT",
          "constrained": [
            false,
            false,
            true
          ],
          "value": [
            0.0,
            0.0,
            0.0
          ]
        }
      },
      {
        "python_module": "assign_vector_variable_process",
        "kratos_module": "KratosMultiphysics",
        "Parameters": {
          "mesh_id": 0,
          "model_part_name": "Structure.DISPLACEMENT_LE",
          "variable_name": "DISPLACEMENT",
          "constrained": [
            false,
            false,
            true
          ],
          "value": [
            0.0,
            0.0,
            0.0
          ]
        }
      }
    ],
    "loads_process_list": [],
    "list_other_processes": []
  },
  "output_processes": {
    "vtk_output": [
      {
        "python_module": "vtk_output_process",
        "kratos_module": "KratosMultiphysics",
        "process_name": "VtkOutputProcess",
        "help": "This process writes postprocessing files for Paraview",
        "Parameters": {
          "model_part_name": "Structure",
          "output_control_type": "step",
          "output_interval": 1,
          "file_format": "ascii",
          "output_precision": 7,
          "output_sub_model_parts": false,
          "output_path": "vtk_output",
          "save_output_files_in_folder": true,
          "nodal_solution_step_data_variables": [
            "DISPLACEMENT",
            "REACTION",
            "POSITIVE_FACE_PRESSURE",
            "REACTION_MOMENT"
          ],
          "nodal_data_value_variables": [],
          "element_data_value_variables": [],
          "condition_data_value_variables": [
            "POSITIVE_FACE_PRESSURE"
          ],
          "gauss_point_variables_extrapolated_to_nodes": [
            "FORCE",
            "MOMENT"
          ]
        }
      }
    ],
    "restart_processes": [
      {
        "python_module": "save_restart_process",
        "kratos_module": "KratosMultiphysics",
        "process_name": "SaveRestartProcess",
        "Parameters": {
          "model_part_name": "Structure",
          "restart_control_type": "step",
          "restart_save_frequency": 10,
          "max_files_to_keep": 1
        }
      }
    ]
  },
  "interface_sub_model_parts_list": [
    "SurfaceLoad3D_hole_master"
  ],
  "pressure_directions": [
    -1
  ]
}

**Comment by loumalouomega** [comment]
Doesn't look wrong, I don't know why is failing. Can you share the example?

**Reply by MicheleBarucca97** [reply]
Did you mean the mdpa  and strcuturalMaterial file?

**Comment by loumalouomega** [comment]
Please attach as a zip

**Reply by loumalouomega** [reply]
I will take a look next week, today I have plenty of work. In any case you can try to use instead of residual criteria displacement criteria, at least if result is not wrong should converge
